exonum/benches/criterion/block.rs

```diff
@@ -0,0 +1,235 @@
+use exonum::storage::{Database, DbOptions, Fork, Patch, ProofMapIndex, RocksDB, Snapshot};
+use futures::sync::mpsc;
+use tempdir::TempDir;
+const TIMESTAMPING_SERVICE_ID: u16 = 1;
+const CRYPTOCURRENCY_SERVICE_ID: u16 = 255;
+
+fn create_blockchain(db: Box<Database>, services: Vec<Box<Service>>) -> Blockchain {
+    let dummy_channel = mpsc::channel(1);
+    let dummy_keypair = (PublicKey::zero(), SecretKey::zero());
+    Blockchain::new(
+        db,
+        services,
+        dummy_keypair.0,
+        dummy_keypair.1,
+        ApiSender::new(dummy_channel.0),
+    )
+}
+
+fn execute_block(blockchain: &Blockchain, height: u64, txs: &[Hash]) -> Patch {
+    blockchain
+        .create_patch(ValidatorId::zero(), Height(height), txs)
+        .1
+}
+
+fn execute_timestamping(db: Box<Database>, c: &mut Criterion) {
+    struct Timestamping;
+
+    impl Service for Timestamping {
+        fn service_id(&self) -> u16 {
+            TIMESTAMPING_SERVICE_ID
+        }
+
+        fn service_name(&self) -> &'static str {
+            "timestamping"
+        }
+
+        fn state_hash(&self, _: &Snapshot) -> Vec<Hash> {
+            Vec::new()
+        }
+
+        fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<Transaction>, EncodingError> {
+            Ok(Box::new(Tx::from_raw(raw)?))
+        }
+    }
+
+    transactions! {
+        TimestampingTransactions {
+            const SERVICE_ID = TIMESTAMPING_SERVICE_ID;
+            struct Tx {
+                from: &PublicKey,
+                data: &Hash,
+            }
+        }
+    }
+
+    impl Transaction for Tx {
+        fn verify(&self) -> bool {
+            self.verify_signature(self.from())
+        }
+
+        fn execute(&self, _: &mut Fork) -> ExecutionResult {
+            Ok(())
+        }
+    }
+
+    fn prepare_txs(blockchain: &mut Blockchain, height: u64, count: u64) -> Vec<Hash> {
+        let mut fork = blockchain.fork();
+        let mut txs = Vec::new();
+        {
+            let mut schema = Schema::new(&mut fork);
+            let (pub_key, sec_key) = gen_keypair();
+            for i in (height * count)..((height + 1) * count) {
+                let tx = Tx::new(&pub_key, &i.hash(), &sec_key);
+                let tx_hash = Transaction::hash(&tx);
+                txs.push(tx_hash);
+                schema.add_transaction_into_pool(tx.raw().clone());
+            }
+        }
+        blockchain.merge(fork.into_patch()).unwrap();
+        txs
+    }
+    let mut blockchain = create_blockchain(db, vec![Box::new(Timestamping)]);
+    for i in 0..100 {
+        let txs = prepare_txs(&mut blockchain, i, 1000);
+        let patch = execute_block(&blockchain, i, &txs);
+        blockchain.merge(patch).unwrap();
+    }
+
+    let txs = prepare_txs(&mut blockchain, 100, 1000);
+
+    c.bench(
+        "timestamping",
+        Benchmark::new("timestamping", move |b| {
+            b.iter(|| execute_block(&blockchain, 100, &txs))
+        }).sample_size(16),
```

Comment:
Nitpick: Why 16 and not another number? I suggest to move out magic numbers to variables or constants with a brief explanation why this value has been chosen and not another. This comments applies also to other magic numbers usages.

Reply:
Because execute_block is not really "micro benchmark"
And execute it as regular benches, with 100 samples, lead to relatively big testing time (10 min on my mac).

This number does not contain any additional information, but is simply a compromise between long lead times and test quality.

Reply:
I guess such a comment will look just great in sources :smiley: